staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -189,15 +189,44 @@ func (t *watcherBookmarkTimeBuckets) addWatcher(w *cacheWatcher) bool {
 	}
 	watchers, _ := t.watchersBuckets[bucketID]
 	t.watchersBuckets[bucketID] = append(watchers, w)
+
+	w.registerForgetBookmark(bucketID, t.clearStopedWatcher)
+
 	return true
 }
 
+func (t *watcherBookmarkTimeBuckets) clearStopedWatcher(bucketID int64) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if watchers, ok := t.watchersBuckets[bucketID]; ok {
+		unstopedWatchers := make([]*cacheWatcher, 0, len(watchers))
+		for _, w := range watchers {
+			if w.stopped {
+				continue
+			}
+			unstopedWatchers = append(unstopedWatchers, w)
+		}
+
+		if len(unstopedWatchers) > 0 {
+			t.watchersBuckets[bucketID] = unstopedWatchers
+		} else {
+			delete(t.watchersBuckets, bucketID)
+		}
+	}
+}
+
 func (t *watcherBookmarkTimeBuckets) popExpiredWatchers() [][]*cacheWatcher {
 	currentBucketID := t.clock.Now().Unix()
 	// There should be one or two elements in almost all cases
 	expiredWatchers := make([][]*cacheWatcher, 0, 2)
 	t.lock.Lock()
 	defer t.lock.Unlock()
+
+	if t.startBucketID > currentBucketID {
```

Comment:
A naive question: should buckets between currentBucketID and t.startBucketID be handled?
By the way, this worth a comment.

Reply:
If we expire buckets between currentBucketID and t.startBucketID when t.startBucketID is bigger than currentBucketID, I think some bookmark pushes maybe failed.
In this modification, the cacheWatcher reference in watchersBuckets will be actively cleaned up when the cacheWatcher is closed. I think this is better.